src/Microsoft.OData.Core/ODataWriterCore.cs

```diff
@@ -1655,7 +1714,10 @@ private void ValidateTransition(WriterState newState)
 
                     break;
                 case WriterState.ResourceSet:
-                    if (newState != WriterState.Resource)
+                    if (newState != WriterState.Resource &&
+                        (this.CurrentScope.ResourceType == null ||
+                            (this.CurrentScope.ResourceType.TypeKind != EdmTypeKind.Untyped ||
+                                (newState != WriterState.Primitive && newState != WriterState.ResourceSet))))

```

Comment:
Added comment:
                    // Within a typed resource set we can only write a resource 
                    // Within an untyped resource set we can also write a primitive value or nested resource set 


---
In reply to: [127299880](https://github.com/OData/odata.net/pull/894#discussion_r127299880) [](ancestors = 127299880)